### PR TITLE
[Feature] Support EarlyStoppingHook

### LINKED ISSUE #356: Support Early Stop
**Describe the feature**

**Motivation**
See an issue in https://github.com/open-mmlab/mmdetection/issues/1383. May stop early stop in mmengine so that all OpenMMLab repos could use this feature.

**Related resources**
If there is an official code release or third-party implementations, please also provide the information here, which would be very helpful.

**Additional context**
If you would like to implement the feature and create a PR, please leave a comment here and that would be much appreciated.

### PR DESCRIPTION
**Pull request by nijkah**
## Motivation
Closes #356 

## Modification
Added a new variable `stop_training` in `EpochBasedTrainLoop` and `IterBasedTrainLoop`.
Edited `run` method as
```python
    def run(self) -> torch.nn.Module:
        """Launch training."""
        while self._epoch < self._max_epochs and not self.stop_training:
            self.run_epoch()
```

**
`EarlyStoppingHook` itself does not save the best checkpoint. Should we support it?

I referred to some logics from
- https://github.com/Bjarten/early-stopping-pytorch/blob/master/pytorchtools.py
- https://docs.ray.io/en/latest/_modules/ray/tune/stopper/experiment_plateau.html#ExperimentPlateauStopper

### CODE REVIEW

File: mmengine/hooks/early_stopping_hook.py
Comment (HAOCHENYE) [inline mmengine/hooks/early_stopping_hook.py:84]: How about using `PriorityQueue`?

File: mmengine/hooks/early_stopping_hook.py
Comment (HAOCHENYE) [inline mmengine/hooks/early_stopping_hook.py:24]: Hi, referring to https://github.com/Bjarten/early-stopping-pytorch/blob/master/pytorchtools.py
Training will be stopped when the metric gets consistently worse "patience" times. The `rule` argument means the better trends we want, and the patience means How long to wait after last time validation metric improved. Such setting makes sense to me.

If I understand correctly, the current implementation will stop training when the metric gets consistently better "patience" times, is it our expected?

Comment (HAOCHENYE) [reply mmengine/hooks/early_stopping_hook.py:24]: BTW, I think we could also adopt the strategy here https://github.com/open-mmlab/mmengine/pull/739#pullrequestreview-1185414869

If we reach the target metric threshold, just stop training.

Comment (nijkah) [reply mmengine/hooks/early_stopping_hook.py:24]: Inspired by [`ray.tune.stopper`](https://docs.ray.io/en/latest/_modules/ray/tune/stopper/experiment_plateau.html#ExperimentPlateauStopper),
my original intention was "Stop training when the current score does not exceed "the delta plus average of top `pool_size` scores. But it is acceptable to `patience` times when the conditions are not met.".

In this aspect, `rule` means that "current score should be `'greater|less'` than the average of latest scores.".
So, to check when the condition does not meet, we put the 'the average of latest scores' as a left argument.
```python
compare(sum(self.pool_values) / self.pool_size + self.delta, latest_value)
```

To my understanding, the current implementation meets your description. Is it better to improve the clarity?

File: mmengine/hooks/early_stopping_hook.py
Comment (nijkah) [reply mmengine/hooks/early_stopping_hook.py:84]: https://stackoverflow.com/a/26441285
If `rule=='greater`, getting values from the `PriorityQueue` should work differently. `Q.get()` or `Q.get()[1]`.
Is there any good idea to handle this case?

File: mmengine/hooks/early_stopping_hook.py
Comment (nijkah) [reply mmengine/hooks/early_stopping_hook.py:69]: Added it!

File: mmengine/hooks/early_stopping_hook.py
Comment (HAOCHENYE) [reply mmengine/hooks/early_stopping_hook.py:94]: Initially, I thought we can put + or - metrics in `PriorityQueue` to represent `greater` or `less`. Emmmm. But I think since the metrics list will not be too long, it could be OK to directly use `list` rather than `PriorityQueue`

File: mmengine/hooks/early_stopping_hook.py
Comment (HAOCHENYE) [reply mmengine/hooks/early_stopping_hook.py:24]: Ah, I don't know what happened, I misunderstood in the morning :laughing: ,my bad :rofl: 

But maybe we should set the default value of patience to 1 or a larger one? I think 0 or 1 is too strict, and the model will be very easy to trigger the early stop.

File: mmengine/hooks/early_stopping_hook.py
Comment (HAOCHENYE) [inline mmengine/hooks/early_stopping_hook.py:16]: Referring to https://keras.io/api/callbacks/early_stopping/, maybe `monitor` will be a more proper parameter name? Maybe someday MMEngine will support calculate the validation loss in the train loop :rofl:

File: mmengine/hooks/early_stopping_hook.py
Comment (HAOCHENYE) [inline mmengine/hooks/early_stopping_hook.py:23]: Ahhh, it could be better to describe the patience in more detailed

File: mmengine/hooks/early_stopping_hook.py
Comment (HAOCHENYE) [inline mmengine/hooks/early_stopping_hook.py:97]: Should delta be "added" or "minors" according to the rule?

File: mmengine/hooks/early_stopping_hook.py
Comment (nijkah) [reply mmengine/hooks/early_stopping_hook.py]: I borrowed the description from https://docs.ray.io/en/latest/_modules/ray/tune/stopper/experiment_plateau.html#ExperimentPlateauStopper.

Is the description from https://keras.io/api/callbacks/early_stopping/ better?

**Comment by nijkah** [comment]
Hi @HAOCHENYE, I have a little question.
In the UT with initializing `Runner`, the error raises when the `experiment_name` of `Runner` isn't specified. Also, it happens when the `experiment_name` is duplicated with the other UTs although there is a logic for cleaning its temporary directory.

Is it intended (the developers should consider) or unintended behavior?

**Comment by HAOCHENYE** [comment]
> Hi @HAOCHENYE, I have a little question. In the UT with initializing `Runner`, the error raises when the `experiment_name` of `Runner` isn't specified. Also, it happens when the `experiment_name` is duplicated with the other UTs although there is a logic for cleaning its temporary directory.
> 
> Is it intended (the developers should consider) or unintended behavior?

Usually, if the user does not specify the experiment name, `Runner` will automatically generate the experiment name and each `Runner` instance will have unique experiment names **with different timestamp suffix**. However, if the time interval between `Runner`'s creation is so short that the timestamp suffixes are identical, it could lead to Runners with duplicate experiment names, which is not allowed in current design.  

MMEngine has multiple global variances such as `MessageHub`,  `Visualizer` and `DefaultScope`, and all these variables are marked with the experiment name of `Runner`. Global variables are very dangerous, therefore these variables must have unique names.

As for the unit tests will often raise an error about the duplicated experiment name, it is caused by the irregularity of the current unit testing, and we plan to refactor the unit tests, and provide a more general function to build Runner in unit tests.

File: mmengine/hooks/early_stopping_hook.py
Comment (HAOCHENYE) [reply mmengine/hooks/early_stopping_hook.py]: LGTM 😆!

File: mmengine/hooks/early_stopping_hook.py
Comment (HAOCHENYE) [reply mmengine/hooks/early_stopping_hook.py]: Yep, I think the description of https://keras.io/api/callbacks/early_stopping/ is better. BTW, the `epoch` in the description is not accurate, will `the times of validation` be better?

File: mmengine/hooks/early_stopping_hook.py
Comment (nijkah) [reply mmengine/hooks/early_stopping_hook.py:97]: I'm not sure about the naming. But, I think `delta` is still intuitive to understand what it means.
Then, how about `min_delta`, which means `minimum difference that must be appeared` as in https://keras.io/api/callbacks/early_stopping/?

**Comment by nijkah** [comment]
One more question! I am thinking about which experiments we should consider for no improvement.

Current implementation retains **top scores** ignoring the recent inferior results.
But it may be more intuitive to retain **latest scores** to check the trend of training.

Which one will be better?

**Comment by HAOCHENYE** [comment]
> One more question! I am thinking about which experiments we should consider for no improvement.
> 
> Current implementation retains **top scores** ignoring the recent inferior results. But it may be more intuitive to retain **latest scores** to check the trend of training.
> 
> Which one will be better?

Additional notes： We could also borrow the implementation in 

https://github.com/Lightning-AI/lightning/blob/9a4e8a8c528d6475ab33c46a0a84e27273cc10bd/src/pytorch_lightning/callbacks/early_stopping.py#L38

Back to the topic, it could make more sense to me to directly compare the latest metric with the previous best Nth metrics and finally update the Nth metrics with the latest one, rather than first update the Nth metrics, and then compare them 🤣. what do you think about it?

**Comment by nijkah** [comment]
I think your provided link looks neat. I'll rewrite the code especially for stopping logic and `__init__` parameters.

File: mmengine/hooks/__init__.py
Comment (zhouzaida) [inline mmengine/hooks/__init__.py:27]:
```suggestion
    'EarlyStoppingHook'
```

File: mmengine/hooks/early_stopping_hook.py
Comment (zhouzaida) [inline mmengine/hooks/early_stopping_hook.py:57]:
```suggestion
            "`rule` should be either 'greater' or 'less'."
```

File: mmengine/hooks/early_stopping_hook.py
Comment (zhouzaida) [inline mmengine/hooks/early_stopping_hook.py:77]:
```suggestion
                                   f'{current_score} is infinite. '
```

File: mmengine/hooks/early_stopping_hook.py
Comment (zhouzaida) [inline mmengine/hooks/early_stopping_hook.py:13]: Consider recording the state of early stopping to support resuming training.

File: mmengine/runner/loops.py
Comment (zhouzaida) [inline mmengine/runner/loops.py:52]:
```suggestion
        # This attribute will be updated by `EarlyStoppingHook`
        # when it is enabled.
        self.stop_training = False
```

File: mmengine/runner/loops.py
Comment (zhouzaida) [inline mmengine/runner/loops.py:220]:
```suggestion
        # This attribute will be updated by `EarlyStoppingHook`
        # when it is enabled.
        self.stop_training = False
```

File: mmengine/hooks/early_stopping_hook.py
Comment (zhouzaida) [inline mmengine/hooks/early_stopping_hook.py:134]:
```suggestion
        if self.monitor not in metrics:
            if self.strict:
                raise RuntimeError(
                    'Early stopping conditioned on metric '
                    f'`{self.monitor} is not available. Please check available'
                    f' metrics {metrics}, or set `strict=False` in '
                    '`EarlyStoppingHook`.')
            warnings.warn(
                'Skip early stopping process since the evaluation '
                f'results ({metrics.keys()}) do not include `monitor` '
                f'({self.monitor}).')
            return
```

**Comment by zhouzaida** [comment]
Please also append the Hook in the following files.

- https://github.com/open-mmlab/mmengine/blob/main/docs/zh_cn/api/hooks.rst
- https://github.com/open-mmlab/mmengine/blob/main/docs/en/api/hooks.rst

File: mmengine/hooks/early_stopping_hook.py
Comment (HAOCHENYE) [inline mmengine/hooks/early_stopping_hook.py:57]:
```suggestion
            '`rule` should be either "greater" or "less".'
```

File: mmengine/hooks/early_stopping_hook.py
Comment (HAOCHENYE) [inline mmengine/hooks/early_stopping_hook.py:67]: Suggest directly return `reason_message` and `stop_training` in `_check_stop_condition`.

File: mmengine/hooks/early_stopping_hook.py
Comment (HAOCHENYE) [reply mmengine/hooks/early_stopping_hook.py:13]: `MessageHub` will save all history metrics during training, maybe we could utilize it to resume training

File: mmengine/hooks/early_stopping_hook.py
Comment (HAOCHENYE) [inline mmengine/hooks/early_stopping_hook.py:67]: `self.default_end_message` does not need to be an attribute anymore.

File: tests/test_hooks/test_early_stopping_hook.py
Comment (HAOCHENYE) [inline tests/test_hooks/test_early_stopping_hook.py:85]: Hi! You can use `RunnerTestCase` defined [here](https://github.com/open-mmlab/mmengine/blob/main/mmengine/testing/runner_test_case.py) to test hooks :smile:

**Review by HAOCHENYE — COMMENTED** [review]
Verrrrrrry thank you for the contributions :laughing: ! This pr will be ready to merge after supporting resuming and updating the unit test :rocket:  :rocket:! Very much looking forward to this new feature!

File: mmengine/hooks/early_stopping_hook.py
Comment (nijkah) [reply mmengine/hooks/early_stopping_hook.py:67]: @HAOCHENYE I'll continue this work this week! Sorry for late response 😢

**Comment by zhouzaida** [comment]
Hi @nijkah , is there any progress?

**Comment by nijkah** [comment]
@zhouzaida Truly apologize for the late response. I think I can start working on it by this Wednesday.

**Comment by zhouzaida** [comment]
Hi @nijkah , is there any progress?

**Comment by nijkah** [comment]
@zhouzaida @HAOCHENYE  Sorry and thanks for reminding me. I'll definitely continue this until through the holiday week.

**Comment by nijkah** [comment]
Hi @HAOCHENYE , really apologize, and thanks for your patience. 
Now, I am thinking about supporting resuming logic.
My initial idea is just to use `save_last` flag and borrow the logic of `CheckpointHook._save_checkpoint(runner)`.

You already mentioned the `MessageHub`, can I ask you to give some guide on how to utilize `MessageHub` to support resuming?

**Comment by HAOCHENYE** [comment]
> Hi @HAOCHENYE , really apologize, and thanks for your patience. Now, I am thinking about supporting resuming logic. My initial idea is just to use `save_last` flag and borrow the logic of `CheckpointHook._save_checkpoint(runner)`.
> 
> You already mentioned the `MessageHub`, can I ask you to give some guide on how to utilize `MessageHub` to support resuming?

Hi! History information is saved in `MessageHub`, such as loss history and metric history. These variables are updated by `MessageHub.update_scalar` like this:

https://github.com/open-mmlab/mmengine/blob/6dc1d7082a1e2bc32049e65f283c4fd20de1c026/mmengine/hooks/runtime_info_hook.py#L120

Then, these variables can be resumed from `runner.message_hub`. Maybe EarlyStoppingHook can refer to this way?

**Comment by zhouzaida** [comment]
Hi @nijkah , thanks for your contribution. We are planning to merge this PR today and release a new version 0.6.0. The feature to resume can be supported in the future if users require it.

File: mmengine/hooks/__init__.py
Comment (zhouzaida) [inline mmengine/hooks/__init__.py:21]:
```suggestion
    'PrepareTTAHook', 'NPUProfilerHook', 'EarlyStoppingHook'
```

File: mmengine/hooks/early_stopping_hook.py
Comment (zhouzaida) [inline mmengine/hooks/early_stopping_hook.py:62]:
```suggestion
        self.min_delta = min_delta if rule == 'greater' else -1 * min_delta
```

File: mmengine/hooks/early_stopping_hook.py
Comment (zhouzaida) [inline mmengine/hooks/early_stopping_hook.py:31]:
```suggestion
            stopping is enabled.
            
       Note:
           `New in version 0.6.0.`
```

File: mmengine/hooks/early_stopping_hook.py
Comment (zhouzaida) [inline mmengine/hooks/early_stopping_hook.py:20]:
```suggestion
        min_delta (float, optional): Minimum difference to continue the
```

**Comment by nijkah** [comment]
> Hi @nijkah , thanks for your contribution. We are planning to merge this PR today and release a new version 0.6.0. The feature to resume can be supported in the future if users require it. In addition, this PR can be merged after resolving the conflict.

@zhouzaida I apologize for not finishing the rest work. I think it is okay to merge right now. I'll try to stay tuned to the future work for this.

**Comment by zhouzaida** [comment]
> > Hi @nijkah , thanks for your contribution. We are planning to merge this PR today and release a new version 0.6.0. The feature to resume can be supported in the future if users require it. In addition, this PR can be merged after resolving the conflict.
> 
> @zhouzaida I apologize for not finishing the rest work. I think it is okay to merge right now. I'll try to stay tuned to the future work for this.

Hi @nijkah , thanks for your contribution again 👍.

File: mmengine/hooks/early_stopping_hook.py
Comment (zhouzaida) [inline mmengine/hooks/early_stopping_hook.py:148]:
```suggestion
                    f'Saving last checkpoint at {runner.epoch} epochs.')
```

Comment (zhouzaida) [reply mmengine/hooks/early_stopping_hook.py:148]: `runner.epoch` will increase by one after training an epoch that is ahead of validation. https://github.com/open-mmlab/mmengine/blob/d8abf9ac0ff053c2fc23f3902edb9aabebbe1ebf/mmengine/runner/loops.py#L109

File: tests/test_hooks/test_early_stopping_hook.py
Comment (zhouzaida) [inline tests/test_hooks/test_early_stopping_hook.py:156]:
```suggestion
        # Check smallest value
```

File: mmengine/hooks/early_stopping_hook.py
Comment (zhouzaida) [inline mmengine/hooks/early_stopping_hook.py:31]: This feature can be removed from the current version because it only prints a message.

File: mmengine/hooks/early_stopping_hook.py
Comment (zhouzaida) [inline mmengine/hooks/early_stopping_hook.py:58]:
```suggestion
        greater_keys = [key.lower() for key in self._default_greater_keys]
        less_keys = [key.lower() for key in self._default_less_keys]
        if monitor.lower() in greater_keys:
            rule = 'greater'
        elif monitor.lower() in less_keys:
            rule = 'less'
```

File: tests/test_hooks/test_early_stopping_hook.py
Comment (zhouzaida) [inline tests/test_hooks/test_early_stopping_hook.py:105]:
```suggestion
        hook = EarlyStoppingHook(monitor='acc')
        assert hook.rule == 'greater'
        assert hook.best_score < 0

        hook = EarlyStoppingHook(monitor='ACC')
        assert hook.rule == 'greater'
        assert hook.best_score < 0        
        

        hook = EarlyStoppingHook(monitor='loss')
        assert hook.rule == 'less'
        assert hook.best_score > 0
        
        hook = EarlyStoppingHook(monitor='Loss')
        assert hook.rule == 'less'
        assert hook.best_score > 0
```

**Review by zhouzaida — APPROVED** [review]

File: tests/test_hooks/test_early_stopping_hook.py
Comment (HAOCHENYE) [inline tests/test_hooks/test_early_stopping_hook.py:100]: Suggest using `self.assertEqual` defined in `unittest.TestCase` to replace with `assert`

File: tests/test_hooks/test_early_stopping_hook.py
Comment (HAOCHENYE) [inline tests/test_hooks/test_early_stopping_hook.py:146]: Suggese using `self.assertWarns
`

File: tests/test_hooks/test_early_stopping_hook.py
Comment (HAOCHENYE) [inline tests/test_hooks/test_early_stopping_hook.py:138]: Suggest using `self.assertRaises`

File: tests/test_hooks/test_early_stopping_hook.py
Comment (HAOCHENYE) [inline tests/test_hooks/test_early_stopping_hook.py:163]: Suggest using self.assertRaises

File: tests/test_hooks/test_early_stopping_hook.py
Comment (HAOCHENYE) [inline tests/test_hooks/test_early_stopping_hook.py:109]: Duplicated?

File: tests/test_hooks/test_early_stopping_hook.py
Comment (HAOCHENYE) [inline tests/test_hooks/test_early_stopping_hook.py:127]: Suggest using `self.assertRaises`

File: tests/test_hooks/test_early_stopping_hook.py
Comment (HAOCHENYE) [inline tests/test_hooks/test_early_stopping_hook.py:130]: Suggest using `self.assertRaises`

Comment (HAOCHENYE) [reply tests/test_hooks/test_early_stopping_hook.py:130]: The rule `the world` reminds me of the comic [jojo](https://jojowiki.com/The_World)(one of my favorite comic :laughing: ). Just missing this comment If I misunderstand it :rofl:

File: tests/test_hooks/test_early_stopping_hook.py
Comment (HAOCHENYE) [inline tests/test_hooks/test_early_stopping_hook.py:198]: `self.assertTrue`

File: tests/test_hooks/test_early_stopping_hook.py
Comment (HAOCHENYE) [inline tests/test_hooks/test_early_stopping_hook.py:209]: Same as above

File: mmengine/hooks/early_stopping_hook.py
Comment (zhouzaida) [inline mmengine/hooks/early_stopping_hook.py:32]:
```suggestion
           `New in version 0.7.0.`
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**
